From a4d901c1670e1cd79650fe58d43d2b93ac3474a0 Mon Sep 17 00:00:00 2001
From: Arko Dasgupta <arkodg@users.noreply.github.com>
Date: Tue, 8 Feb 2022 05:47:30 +0530
Subject: Fix IOP merge for ingressgategateway label field (#37191)

* Fix IOP merge for ingressgategateway label field

Fixes: https://github.com/istio/istio/issues/37190

Signed-off-by: Arko Dasgupta <arko@tetrate.io>

* fix testdata output

Signed-off-by: Arko Dasgupta <arko@tetrate.io>
---
 operator/pkg/util/merge_iop.go                    |  3 ++-
 operator/pkg/util/merge_iop_test.go               | 15 +++++++++++++++
 .../pkg/util/testdata/yaml/input/yaml_layer1.yaml |  2 ++
 .../pkg/util/testdata/yaml/input/yaml_layer2.yaml |  2 ++
 .../pkg/util/testdata/yaml/output/layer1.yaml     |  2 ++
 .../pkg/util/testdata/yaml/output/layer1_2.yaml   |  3 +++
 .../pkg/util/testdata/yaml/output/layer1_2_3.yaml |  3 +++
 .../util/testdata/yaml/output/layer1_stdin.yaml   |  2 ++
 8 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/operator/pkg/util/merge_iop.go b/operator/pkg/util/merge_iop.go
index ee70d4dcdb..e39460a5c8 100644
--- a/operator/pkg/util/merge_iop.go
+++ b/operator/pkg/util/merge_iop.go
@@ -64,7 +64,8 @@ type componentSpec struct {
 }
 
 type gatewaySpec struct {
-	K8S *v1alpha1.KubernetesResourcesSpec `json:"k8s" patchStrategy:"merge"`
+	Label map[string]string                 `json:"label" patchStrategy:"merge"`
+	K8S   *v1alpha1.KubernetesResourcesSpec `json:"k8s" patchStrategy:"merge"`
 }
 
 type values struct {
diff --git a/operator/pkg/util/merge_iop_test.go b/operator/pkg/util/merge_iop_test.go
index 05fc533573..962fd7871c 100644
--- a/operator/pkg/util/merge_iop_test.go
+++ b/operator/pkg/util/merge_iop_test.go
@@ -65,3 +65,18 @@ func TestOverlayIOPDefaultMeshConfig(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestOverlayIOPIngressGatewayLabel(t *testing.T) {
+	l1, err := os.ReadFile("testdata/yaml/input/yaml_layer1.yaml")
+	if err != nil {
+		t.Fatal(err)
+	}
+	l2, err := os.ReadFile("testdata/yaml/input/yaml_layer2.yaml")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := OverlayIOP(string(l1), string(l2)); err != nil {
+		t.Fatal(err)
+	}
+}
diff --git a/operator/pkg/util/testdata/yaml/input/yaml_layer1.yaml b/operator/pkg/util/testdata/yaml/input/yaml_layer1.yaml
index 76146a74fc..d35f837e76 100644
--- a/operator/pkg/util/testdata/yaml/input/yaml_layer1.yaml
+++ b/operator/pkg/util/testdata/yaml/input/yaml_layer1.yaml
@@ -10,6 +10,8 @@ spec:
       - namespace: istio-system
         name: istio-ingressgateway
         enabled: true
+        label:
+          api: default
         k8s:
           service:
             externalTrafficPolicy: Local
diff --git a/operator/pkg/util/testdata/yaml/input/yaml_layer2.yaml b/operator/pkg/util/testdata/yaml/input/yaml_layer2.yaml
index 96540e0a5b..bde7b1921a 100644
--- a/operator/pkg/util/testdata/yaml/input/yaml_layer2.yaml
+++ b/operator/pkg/util/testdata/yaml/input/yaml_layer2.yaml
@@ -8,6 +8,8 @@ spec:
       - namespace: istio-system
         name: istio-ingressgateway
         enabled: true
+        label:
+          foo: bar
         k8s:
           service:
             externalTrafficPolicy: Test
diff --git a/operator/pkg/util/testdata/yaml/output/layer1.yaml b/operator/pkg/util/testdata/yaml/output/layer1.yaml
index 76146a74fc..d35f837e76 100644
--- a/operator/pkg/util/testdata/yaml/output/layer1.yaml
+++ b/operator/pkg/util/testdata/yaml/output/layer1.yaml
@@ -10,6 +10,8 @@ spec:
       - namespace: istio-system
         name: istio-ingressgateway
         enabled: true
+        label:
+          api: default
         k8s:
           service:
             externalTrafficPolicy: Local
diff --git a/operator/pkg/util/testdata/yaml/output/layer1_2.yaml b/operator/pkg/util/testdata/yaml/output/layer1_2.yaml
index c268921db6..1820a14275 100644
--- a/operator/pkg/util/testdata/yaml/output/layer1_2.yaml
+++ b/operator/pkg/util/testdata/yaml/output/layer1_2.yaml
@@ -6,6 +6,9 @@ spec:
       enabled: false
     ingressGateways:
       - enabled: true
+        label:
+          api: default
+          foo: bar
         k8s:
           service:
             externalTrafficPolicy: Test
diff --git a/operator/pkg/util/testdata/yaml/output/layer1_2_3.yaml b/operator/pkg/util/testdata/yaml/output/layer1_2_3.yaml
index ce06f30b5b..b3ec6be7b7 100644
--- a/operator/pkg/util/testdata/yaml/output/layer1_2_3.yaml
+++ b/operator/pkg/util/testdata/yaml/output/layer1_2_3.yaml
@@ -6,6 +6,9 @@ spec:
       enabled: true
     ingressGateways:
       - enabled: false
+        label:
+          api: default
+          foo: bar
         k8s:
           service:
             externalTrafficPolicy: Test
diff --git a/operator/pkg/util/testdata/yaml/output/layer1_stdin.yaml b/operator/pkg/util/testdata/yaml/output/layer1_stdin.yaml
index 76146a74fc..d35f837e76 100644
--- a/operator/pkg/util/testdata/yaml/output/layer1_stdin.yaml
+++ b/operator/pkg/util/testdata/yaml/output/layer1_stdin.yaml
@@ -10,6 +10,8 @@ spec:
       - namespace: istio-system
         name: istio-ingressgateway
         enabled: true
+        label:
+          api: default
         k8s:
           service:
             externalTrafficPolicy: Local
-- 
2.35.3

